### PR TITLE
HIBERNATE-87: Support HQL string contatenation

### PR DESCRIPTION
**Pull request by apmasell**
Adds HQL functions for most string operations, including concatenation. Also, build out infrastructure for binding Mongo operators as functions.

### CODE REVIEW

**Comment by jyemin** [comment]
@apmasell I'm going to wait until you resolve all the Copilot-initiated comment threads before reviewing.

File: src/main/java/com/mongodb/hibernate/internal/translate/mongoast/AstNamedOperatorExpression.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * A general-purpose Mongo operator that includes arguments as a document
+ *
+ * @param operator the name of the operator, including the <code>$</code>
+ * @param arguments a map of name-argument pairs that will be converted into a document
+ */
+public record AstNamedOperatorExpression(String operator, Map<String, AstExpression> arguments)
+        implements AstExpression {
+    @Override
```
[the remark this answers is not in the document]
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/translate/mongoast/AstNamedOperatorExpression.java]: Not relevant.

**Review by jyemin — COMMENTED** [review]
There are a few bugs in this PR related to parameter binder-related defects.  This is at least the third time our parameter binding design has bit us, and it points to bad design of the translator.  Rather than working around it yet again, I've proposed a fix to the design in https://github.com/mongodb/mongo-hibernate/pull/205.  If we merge that PR and then merge this one on top of it, it should address all the issues and we won't have to think about this source of bugs again.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    @Override
+    public void render(
+            SqlAppender sqlAppender,
+            List<? extends SqlAstNode> arguments,
+            ReturnableType<?> returnType,
+            SqlAstTranslator<?> walker) {
+        var translator = AbstractMqlTranslator.cast(walker);
+        var padding = arguments.size() < 3
+                ? new AstLiteralExpression(new AstLiteral(new BsonString(" ")))
+                : new AstUnaryOperatorExpression("$toString", translator.acceptAndYield(arguments.get(2), EXPRESSION));
```
Comment (jyemin) [inline src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java]: This is the one function (besides concat) that coerces with "$toString". For consistency, let's remove it.  Hibernate will already prevent this function from being called on anything but a string anyway.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                                typeConfiguration, FunctionParameterType.STRING),
+                        StandardFunctionArgumentTypeResolvers.impliedOrInvariant(
+                                typeConfiguration, FunctionParameterType.INTEGER),
+                        StandardFunctionArgumentTypeResolvers.impliedOrInvariant(
+                                typeConfiguration, FunctionParameterType.STRING)));
+        this.left = left;
+    }
+
+    @Override
+    public void render(
```
Comment (jyemin) [inline src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java:80]: There are a few gotchas with `pad`.

1. `length < 1` should evaluate to empty string
2. `length < lenth(matchedString)` should truncate `matchedString` from the right 

So these tests should pass

```
@Test
    void testPadLeadingTruncatesWhenTargetIsShorterThanInput() {
        assertResult("select pad(s with 3 leading) from Item", " He");
    }

    @Test
    void testPadTrailingTruncatesWhenTargetIsShorterThanInput() {
        assertResult("select pad(s with 3 trailing) from Item", " He");
    }

    @Test
    void testPadWithZeroTargetReturnsEmptyString() {
        assertResult("select pad(s with 0 leading) from Item", "");
    }

    @Test
    void testPadWithNegativeTargetReturnsEmptyString() {
        assertResult("select pad(s with -1 leading) from Item", "");
    }
```

File: src/main/java/com/mongodb/hibernate/internal/dialect/MongoDialect.java [hunk 4]
```diff
@@ -284,12 +303,85 @@ public void initializeFunctionRegistry(FunctionContributions functionContributio
...
+                        "locate",
+                        "$indexOfCP",
+                        typeConfiguration,
+                        StandardBasicTypes.INTEGER,
+                        FunctionParameterDefinition::addOne,
+                        allModifications(swap(0, 1), offsetToEnd(2, 3)),
+                        required(FunctionParameterType.STRING),
+                        required(FunctionParameterType.STRING),
+                        orMissing(FunctionParameterType.INTEGER).map(FunctionParameterDefinition::subtractOne),
+                        orMissing(FunctionParameterType.INTEGER)));
```
Comment (jyemin) [inline src/main/java/com/mongodb/hibernate/internal/dialect/MongoDialect.java]: I think we should stick to Hibernate's 2 or 3 argument `locate` function rather than add our own 4th argument.

File: src/integrationTest/java/com/mongodb/hibernate/query/function/string/StringFunctionIntegrationTests.java [hunk 5]
```diff
@@ -0,0 +1,943 @@
...
+                        }
+                      }
+                    }
+                  ]
+                }
+                """);
+    }
+
+    @Test
+    void testLocateOffset() {
```
Comment (jyemin) [inline src/integrationTest/java/com/mongodb/hibernate/query/function/string/StringFunctionIntegrationTests.java]: Let's test the three-argument form as well (regardless of whether we remove the four-argument form).

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+import org.hibernate.type.StandardBasicTypes;
+import org.hibernate.type.spi.TypeConfiguration;
+
+public final class MongoTrimFunction extends AbstractSqmSelfRenderingFunctionDescriptor implements ExpressionFunction {
+
+    public MongoTrimFunction(TypeConfiguration typeConfiguration) {
+        super(
+                "trim",
+                new ArgumentTypesValidator(
+                        StandardArgumentsValidators.between(2, 3),
```
Comment (jyemin) [inline src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java]: This can be `exactly(3)` since Hibernate guarantees it.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    @Override
+    public void render(
+            SqlAppender sqlAppender,
+            List<? extends SqlAstNode> arguments,
+            ReturnableType<?> returnType,
+            SqlAstTranslator<?> walker) {
+        var translator = AbstractMqlTranslator.cast(walker);
+        if (arguments.size() != 2 && arguments.size() != 3) {
```
Comment (jyemin) [inline src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java]: By our convention, this should be an assertion since Hibernate should guarantee it, e.g. `assertTrue(arguments.size == 3)`.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+            throw new IllegalArgumentException("Invalid number of arguments to trim");
+        }
+        final var namedArguments = new TreeMap<String, AstExpression>();
```
Comment (jyemin) [inline src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java]: Nit: remove `final`

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        namedArguments.put("input", translator.acceptAndYield(arguments.get(arguments.size() - 1), EXPRESSION));
```
Comment (jyemin) [inline src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java]: Hibernate guarantees three arguments so this can become `arguments.get(2)`

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+        if (arguments.size() == 3) {
```
Comment (jyemin) [inline src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java]: The conditional can be removed since Hibernate guarantees three args.

File: src/main/java/com/mongodb/hibernate/internal/dialect/MongoDialect.java [hunk 11]
```diff
@@ -284,12 +303,85 @@ public void initializeFunctionRegistry(FunctionContributions functionContributio
...
+        functionRegistry.register("array_list", new MongoArrayConstructorFunction(true));
+        functionRegistry.register(
+                "character_length",
+                new MongoExpressionUnaryFunction(
+                        "character_length",
+                        "$strLenCP",
+                        typeConfiguration,
+                        StandardBasicTypes.INTEGER,
+                        FunctionParameterType.STRING));
+        functionRegistry.register(
```
Comment (jyemin) [inline src/main/java/com/mongodb/hibernate/internal/dialect/MongoDialect.java:316]: Let's require a minimum of 1 argument to match Hibernate's existing behavior.

File: src/integrationTest/java/com/mongodb/hibernate/query/function/string/StringFunctionIntegrationTests.java [hunk 12]
```diff
@@ -0,0 +1,943 @@
...
+                                  }
+                                }
+                              }
+                            }
+                          ]
+                        }
+                """);
+    }
+
+    @Test
```
Comment (jyemin) [inline src/integrationTest/java/com/mongodb/hibernate/query/function/string/StringFunctionIntegrationTests.java:2626]: There's some edge cases that need to be accounted for with substring: SQL standard defines the result as the characters whose positions fall in [start, start + len) intersected with [1, length], so a start below 1 is (unfortunately, perhaps) legal and clips the length.

```
    @Test
    void testSubstringWithZeroStart() {
        assertResult("select substring(s, 0) from Item", " Hello ");
    }

    @Test
    void testSubstringFromNegativeStartForLength() {
        assertResult("select substring(s -1, 3) from Item", " ");
    }

    @Test
    void testSubstringWithZeroStartAndLength() {
        assertResult("select substring(s, 0, 3) from Item", " H");
    }
```

File: src/main/java/com/mongodb/hibernate/internal/dialect/MongoDialect.java [hunk 13]
```diff
@@ -284,12 +303,85 @@ public void initializeFunctionRegistry(FunctionContributions functionContributio
...
+                        orMissing(FunctionParameterType.INTEGER)));
+        functionRegistry.register(
+                "lower",
+                new MongoExpressionUnaryFunction(
+                        "lower",
+                        "$toLower",
+                        typeConfiguration,
+                        StandardBasicTypes.STRING,
+                        FunctionParameterType.STRING));
+        functionRegistry.register("lpad", new MongoPadFunction(typeConfiguration, true));
```
Comment (jyemin) [inline src/main/java/com/mongodb/hibernate/internal/dialect/MongoDialect.java:348]: Let's not support `lpad` or `rpad`, with all their complications, and instead just support `pad`, which only allows a single-character padding string.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoExpressionNamedFunction.java [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    @Override
+    public void render(
+            SqlAppender sqlAppender,
+            List<? extends SqlAstNode> arguments,
+            ReturnableType<?> returnType,
+            SqlAstTranslator<?> walker) {
+        var translator = AbstractMqlTranslator.cast(walker);
+        final var namedArguments = new TreeMap<String, AstExpression>();
```
Comment (jyemin) [inline src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoExpressionNamedFunction.java]: Let's kill all the `final` on local variables.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java [hunk 15]
```diff
@@ -0,0 +1,469 @@
...
+
+    static final class PositionalCheck implements RelativePositionChecker {
+        private int required;
+        private int withDefault;
+        private int missing;
+
+        @Override
+        public void incrementRequired() {
+            if (withDefault > 0 || missing > 0) {
+                throw new IllegalArgumentException(
```
Comment (jyemin) [inline src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java]: I think by convention all the `IllegalArgumentException` in this PR should instead be assertions, via `MongoAssertions` methods.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java [hunk 16]
```diff
@@ -0,0 +1,469 @@
...
+     * Define an optional parameter that will have a default value inserted if not provided in the HQL
+     *
+     * <p>Note that the type of this parameter is inferred to match the type of the default value provided.
+     *
+     * @param name the property name that should be emitted in the MQL
+     * @param defaultValue the default value to insert if a matching argument is not supplied in the HQL
+     * @return a parameter definition
+     * @see MongoExpressionNamedFunction
+     */
+    public static FunctionParameterDefinition<String> orDefault(String name, BsonValue defaultValue) {
```
Comment (jyemin) [inline src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java]: This method is currently unused.  Let's delete it and add it later if it becomes necessary.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java [hunk 17]
```diff
@@ -0,0 +1,469 @@
...
+
+    /**
+     * Define an optional parameter that will be omitted if not provided in the HQL
+     *
+     * @param type the type of the parameter
+     * @return a parameter definition
+     * @see MongoExpressionNamedFunction
+     */
+    @SuppressWarnings("NullAway")
+    public static Missing<Void> orMissing(FunctionParameterType type) {
```
Comment (jyemin) [inline src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java]: Let's change both `orMissing` methods to return `FunctionParameterDefinition<Void>` and then make the `Missing` class package-private to match the rest of the `FunctionParameterDefinition` implementations.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java [hunk 18]
```diff
@@ -0,0 +1,469 @@
...
+    @SuppressWarnings("varargs")
+    static <N> FunctionArgumentTypeResolver typeResolverFromParameters(
+            TypeConfiguration typeConfiguration, FunctionParameterDefinition<N>... parameters) {
+        return StandardFunctionArgumentTypeResolvers.byArgument(Stream.of(parameters)
+                .map(parameter ->
+                        StandardFunctionArgumentTypeResolvers.impliedOrInvariant(typeConfiguration, parameter.type()))
+                .toArray(FunctionArgumentTypeResolver[]::new));
+    }
+
+    @SuppressWarnings("varargs")
```
Comment (jyemin) [inline src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java]: Looks like this annotation is no longer needed.

File: src/main/java/com/mongodb/hibernate/internal/translate/mongoast/AstLetBindingExpression.java [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+
+package com.mongodb.hibernate.internal.translate.mongoast;
+
+import java.util.SortedMap;
+import java.util.function.Consumer;
+import org.bson.BsonWriter;
+import org.hibernate.sql.exec.spi.JdbcParameterBinder;
+
+/** @hidden */
+@SuppressWarnings("MissingSummary")
```
Comment (jyemin) [inline src/main/java/com/mongodb/hibernate/internal/translate/mongoast/AstLetBindingExpression.java]: Let's be consistent and add a summary, as in the other two new expressions.

**Review by jyemin — CHANGES_REQUESTED** [review]

File: src/integrationTest/java/com/mongodb/hibernate/query/function/string/StringFunctionIntegrationTests.java (hunk 12 above)
Comment (apmasell) [reply src/integrationTest/java/com/mongodb/hibernate/query/function/string/StringFunctionIntegrationTests.java:2626]: Added.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java (hunk 2 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java]: Removed.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java (hunk 3 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java:80]: Tests added and implementation updated.

File: src/main/java/com/mongodb/hibernate/internal/dialect/MongoDialect.java (hunk 4 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/MongoDialect.java]: Updated.

File: src/integrationTest/java/com/mongodb/hibernate/query/function/string/StringFunctionIntegrationTests.java (hunk 5 above)
Comment (apmasell) [reply src/integrationTest/java/com/mongodb/hibernate/query/function/string/StringFunctionIntegrationTests.java]: Added.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java (hunk 6 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java]: Changed.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java (hunk 7 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java]: This entire check isn't required and I've removed it.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java (hunk 8 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java]: I think I got all of them.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java (hunk 9 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java]: Fixed.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java (hunk 10 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoTrimFunction.java]: Removed.

File: src/main/java/com/mongodb/hibernate/internal/dialect/MongoDialect.java (hunk 11 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/MongoDialect.java:316]: Updated in the variadic function.

File: src/main/java/com/mongodb/hibernate/internal/dialect/MongoDialect.java (hunk 13 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/MongoDialect.java:348]: I think I've ironed out the inconsistencies. Since Hibernate expects to delegate to these functions in the other dialects, I'd rather keep the separate implementations.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoExpressionNamedFunction.java (hunk 14 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoExpressionNamedFunction.java]: Fixed.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java (hunk 15 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java]: I've removed most of these. I can't remove it from the `switch` because the assertion method returns void.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java (hunk 16 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java]: Removed.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java (hunk 17 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java]: Updated.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java (hunk 18 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java]: Removed.

File: src/main/java/com/mongodb/hibernate/internal/translate/mongoast/AstLetBindingExpression.java (hunk 19 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/translate/mongoast/AstLetBindingExpression.java]: Added.

File: src/main/java/module-info.java [hunk 20]
```diff
@@ -117,6 +117,7 @@
     requires transitive org.mongodb.driver.core;
     requires transitive org.mongodb.driver.sync.client;
     requires org.jspecify;
+    requires jdk.compiler;
```
Comment (jyemin) [inline src/main/java/module-info.java]: Not sure how this slipped in but it can be removed.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java (hunk 2 above)
Comment (jyemin) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java]: Still seeing the toString here.  Did you remove a different one?

File: src/main/java/com/mongodb/hibernate/internal/dialect/MongoDialect.java (hunk 13 above)
Comment (jyemin) [reply src/main/java/com/mongodb/hibernate/internal/dialect/MongoDialect.java:348]: Sounds good.

File: src/integrationTest/java/com/mongodb/hibernate/query/function/string/StringFunctionIntegrationTests.java (hunk 12 above)
Comment (jyemin) [reply src/integrationTest/java/com/mongodb/hibernate/query/function/string/StringFunctionIntegrationTests.java:2626]: Another edge case: `adjustedStart` clamps the start, but the length is not floored, so a start far enough below 1 makes the third argument to `$substrCP` negative. On `s = " Hello "`:

| HQL | result |
|---|---|
| `substring(s, -5, 2)` | `Location34454`, `$substrCP: length must be a nonnegative integer` |
| `substring(s, -10, 2)` | same |
| `substring(s, :st, :ln)` with -5 and 2 | same |
| `substring(s, <int field> - 10, 2)` | same |

`substring(s, -1, 3)` returns `" "` by the rule in my earlier comment, and that same rule makes all four of these `""`.

Please floor the length at 0, so that the `length` local emits

```
{"$max": [{"$add": ["$$len", {"$subtract": ["$$start", "$$adjustedStart"]}]}, 0]}
```

File: src/main/java/com/mongodb/hibernate/internal/dialect/MongoDialect.java [hunk 21]
```diff
@@ -284,12 +300,75 @@ public void initializeFunctionRegistry(FunctionContributions functionContributio
...
+                new MongoExpressionPositionalFunction(
+                        "locate",
+                        "$indexOfCP",
+                        typeConfiguration,
+                        StandardBasicTypes.INTEGER,
+                        FunctionParameterDefinition::addOne,
+                        swap(0, 1),
+                        required(FunctionParameterType.STRING),
+                        required(FunctionParameterType.STRING),
+                        orMissing(FunctionParameterType.INTEGER).map(FunctionParameterDefinition::subtractOne)));
```
Comment (jyemin) [inline src/main/java/com/mongodb/hibernate/internal/dialect/MongoDialect.java]: `locate` subtracts one from the start and hands it straight to `$indexOfCP`, so a start below 1 reaches the server: `locate('o', s, 0)` gives `Location40097`, `$indexOfCP requires a nonnegative starting index, found: -1`.

Please floor that argument at 0, so the third element of `$indexOfCP` becomes `{"$max": [{"$subtract": [...]}, 0]}`.   Something like:

```
orMissing(FunctionParameterType.INTEGER)
          .map(FunctionParameterDefinition::subtractOne)
          .map(FunctionParameterDefinition::atLeastZero)
```

Test cases: `locate('o', s, 0)` and `locate('o', s, -3)` both give 6, while `locate(' ', s, 4)` stays 7

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java (hunk 3 above)
Comment (jyemin) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java:80]: All eight pad tests emit one of four pipelines: lpad and rpad crossed with default and explicit padding. The `$cond` tree is emitted whole every time and its arms are selected by the server from the three values in `vars`, so the open coverage question is which arms execute.

| arm | lpad | rpad |
|---|---|---|
| `targetLen <= 0` | `pad(s with -1 leading)` | missing |
| `strLenCP(baseStr) >= targetLen` | `pad(s with 3 leading)` | missing |
| `strLenCP(padding) == 0` | `lpad(s, 60, '')` | missing |
| pad, single-character padding | `with 20 leading[ '*']` | `with 20 trailing[ '*']` |
| pad, multi-character, non-exact division | `lpad('x', 6, 'ab')` | missing |
| pad, multi-character, exact division | missing | missing |
| pad, padding longer than the gap | missing | missing |

Please cover the missing arms. Results on `s = " Hello "`:

| HQL | result |
|---|---|
| `pad(s with -1 trailing)` | `""` |
| `pad(s with 0 leading)` | `""` |
| `pad(s with 3 trailing)` | `" He"` |
| `rpad(s, 60, '')` | `" Hello "` |
| `rpad('abc', 6, 'ab')` | `"abcaba"` |
| `lpad('ab', 6, 'ab')` | `"ababab"` |
| `rpad('ab', 6, 'ab')` | `"ababab"` |
| `lpad('x', 3, 'abcdef')` | `"abx"` |
| `rpad('x', 3, 'abcdef')` | `"xab"` |

These all emit a pipeline the existing tests already asserts, so we should try to avoid duplicating the pipelines in every new test. I suggest adding a `@ParameterizedTest` per pipeline shape, with one row per arm. `pad(s with N leading 'c')` and `lpad(a, n, p)` resolve to the same descriptor, so rows can mix both forms, which also covers `rpad` by name for the first time. 

`MongoSubstringFunction` has the same structure, two pipelines with the `adjustedStart` `$cond` chosen at runtime, so the clamping cases for `substring` fit the same treatment.

**Review by jyemin — CHANGES_REQUESTED** [review]

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java (hunk 2 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java]: Oops. Yes. I think I got them all.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java (hunk 3 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java:80]: I've added the tests, but I'm unclear what you're asking for with `@ParameterizedTest`. Are you looking to have each test check both `pad` and `lpad`/`rpad` with the same paramters?

File: src/main/java/module-info.java (hunk 20 above)
Comment (apmasell) [reply src/main/java/module-info.java]: Removed.

File: src/main/java/com/mongodb/hibernate/internal/dialect/MongoDialect.java (hunk 21 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/MongoDialect.java]: Added.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java (hunk 3 above)
Comment (jyemin) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java:80]: Here's an example of what I mean:

```java

    @ParameterizedTest
    @CsvSource({
        "' ', 4, 7",
        "'o', 0, 6",
        "'o', -3, 6",
    })
    void testLocateWithStart(String pattern, int start, int expected) {
        assertQueryResult(
                "select locate('%s', s, %d) from Item".formatted(pattern, start),
                expected,
                """
                {
                  "aggregate": "items",
                  "pipeline": [
                    {
                      "$project": {
                        "#c_1": {
                          "$add": [
                            {
                              "$indexOfCP": [
                                "$s",
                                "%s",
                                {
                                  "$let": {
                                    "in": {
                                      "$cond": [
                                        {
                                          "$lt": [
                                            "$$x",
                                            {
                                              "$literal": 0
                                            }
                                          ]
                                        },
                                        {
                                          "$literal": 0
                                        },
                                        "$$x"
                                      ]
                                    },
                                    "vars": {
                                      "x": {
                                        "$subtract": [
                                          %d,
                                          {
                                            "$literal": 1
                                          }
                                        ]
                                      }
                                    }
                                  }
                                }
                              ]
                            },
                            {
                              "$literal": 1
                            }
                          ]
                        }
                      }
                    }
                  ]
                }
                """
                        .formatted(pattern, start));

```

The same transform applies to the 16 pad tests: They emit four pipelines, lpad and rpad crossed with default and explicit padding, so they reduce to four parameterized tests:

| pipeline | current tests |
|---|---|
| lpad, default padding | `testPadLeadingWS`, `testPadLeadingTruncatesWhenTargetIsShorterThanInput`, `testPadLeadingWithNegativeTargetReturnsEmptyString` |
| lpad, explicit padding | `testPadLeadingCh`, `testLpadPadsToRequestedLengthWithMultiCharacterPadding`, `testLpadWithEmptyPaddingReturnsInput`, `testLpadMuliCharExact`, `testLpadLongPadding` |
| rpad, default padding | `testPadTrailingWS`, `testPadTrailingTruncatesWhenTargetIsShorterThanInput`, `testPadTrailingWithNegativeTargetReturnsEmptyString` |
| rpad, explicit padding | `testPadTrailingCh`, `testRpadPadsToRequestedLengthWithMultiCharacterPadding`, `testRpadWithEmptyPaddingReturnsInput`, `testRpadMuliCharExact`, `testRpadLongPadding` |

The four three-argument substring tests, `testSubstringLen`, `testSubstringFromNegativeStartForLength`, `testSubstringWithZeroStartAndLength` and `testSubstringNegative`, reduce the same way.

Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java:80]: That is much more difficult to read, follow, and maintain.

Comment (jyemin) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/MongoPadFunction.java:80]: I don't agree, but it's a judgement call.  If you want to keep it like it is, then I will abide by that.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java [hunk 22]
```diff
@@ -223,6 +215,27 @@ public final FunctionParameterDefinition<N> map(Function<? super AstExpression,
         return new Mapped<>(this, mapper);
     }
 
+    /**
+     * Ensures that an input value, which must be a number, is greater than zero
+     *
+     * @param input the expression to clamp
+     * @return an expression which returns the original value or zero if the original value is less than zero
+     */
+    public static AstExpression atLeastZero(AstExpression input) {
```
Comment (jyemin) [inline src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java:219]: Let's replace this with [$max](https://www.mongodb.com/docs/manual/reference/operator/aggregation/max/#use-in-project-stage), like. 

```
{"$max": [<input>, 0]}
```

**Review by jyemin — CHANGES_REQUESTED** [review]
One more simplification request.

File: src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java (hunk 22 above)
Comment (apmasell) [reply src/main/java/com/mongodb/hibernate/internal/dialect/function/FunctionParameterDefinition.java:219]: Fixed.

**Review by jyemin — APPROVED** [review]
LGTM

File: src/integrationTest/java/com/mongodb/hibernate/query/function/string/StringFunctionIntegrationTests.java (hunk 12 above)
Comment (jyemin) [reply src/integrationTest/java/com/mongodb/hibernate/query/function/string/StringFunctionIntegrationTests.java:2626]: Sorry, just noticed this one has no response yet.

Comment (apmasell) [reply src/integrationTest/java/com/mongodb/hibernate/query/function/string/StringFunctionIntegrationTests.java:2626]: This got an `atLeastZero`.
